Add support for basic proto types

Aniket21mathur opened the pull request:
Add support for other proto field types- DOUBLE, FLOAT, UINT64, FIXED64, FIXED32, BOOL , STRING , BYTES, UINT32, SFIXED32, SFIXED64, SINT32, SINT64.

Src file -> https://github.com/Aniket21mathur/ProtocolBuffers/blob/master/chapel/src/types.proto
Output file-> https://github.com/Aniket21mathur/ProtocolBuffers/blob/master/chapel/build/types.chpl

mppf commented on src/Encoding.chpl:
```diff
@@ -0,0 +1,255 @@
+      shiftVal = newVal >> 7;
+      var k = if shiftVal != 0 then 0x80 else 0x00;
+      var newByte = (newVal & 0x7F | k):uint(8);
+      ch.writeBytes(newByte, 1);
```
This does not seem write to me. Anyway in general I don't think this code should be calling readBytes / writeBytes; rather the channel should be configured for binary I/O (and specify the appropriate endianness) and then this code should just do read and write.

Aniket21mathur replied:
I guess you are pointing towards something like this ->
```
var f = open("out", iomode.cw, style = new iostyle(binary=1));
```
If our current implementation is seen, the channel passed to the serialize/unserialize procs is controlled by the user. Is there any other way we can achieve the binary configuration?

mppf replied:
Yes, there are two ways.

First, one can call 'set_style' - see https://github.com/chapel-lang/chapel/blob/c54a5da1fec9f5d889a805c54485d728db6dc467/modules/standard/IO.chpl#L2443 and https://github.com/chapel-lang/chapel/blob/c54a5da1fec9f5d889a805c54485d728db6dc467/modules/standard/IO.chpl#L2459 . You will want to lock the channel before doing these (and unlock it after) - see https://github.com/chapel-lang/chapel/blob/c54a5da1fec9f5d889a805c54485d728db6dc467/modules/standard/IO.chpl#L2193 . I think it would make sense to lock the channel for the entire serialization / deserialization operation.

Second, it's possible to create a new channel based upon another existing channel. For example see https://github.com/chapel-lang/chapel/blob/c54a5da1fec9f5d889a805c54485d728db6dc467/modules/standard/IO.chpl#L3162-L3169 . This is relevant because the Chapel channel types incudes a `param` (known at compile time) indicating binary I/O and endianness (see code starting at https://github.com/chapel-lang/chapel/blob/c54a5da1fec9f5d889a805c54485d728db6dc467/modules/standard/IO.chpl#L461 ) . So, it is possible to wrap the existing channel in a binary-big-endian channel, for example. This can be faster and allows us to skip even code handling locking. However there isn't a good user-facing way to do it yet (although I think such a thing would be reasonable to add to the I/O system; if we choose this path, the protobuf code will use some internal details of channel - or we have to add the user-facing thing to IO.chpl soon).

mppf replied:
Oh, another interesting thing about the second option, is that it would allow all of your functions working with channels to have a concrete type.

mppf replied:
A third option is to use `writef`/`readf`. In practice this will just modify the channel's style to indicate binary I/O for each operation. (You could do that also - modify the style per-operation).

Aniket21mathur commented:
@mppf @astatide @lydia-duncan I have incorporated the binary channel approach, we can use the user exposed function directly once https://github.com/chapel-lang/chapel/pull/16032 is merged. I have also done renaming of user exposed/ library functions, and removed the unused `fieldNumber` argument. I have updated the output file in the Pr message.

Aniket21mathur commented on src/Encoding.chpl:
```diff
@@ -0,0 +1,254 @@
+    }
+  }
+
+  proc unsignedVarintConsume(ch): (uint, int) throws {
```
@mppf I think now we can make the `ch` argument concrete? Something like this ->

`proc unsignedVarintConsume(ch: channel(false,little,false)): (uint, int) throws {`

mppf replied:
Yes

Aniket21mathur replied:
Yup did that :+1:

mppf commented on src/Encoding.chpl:
```diff
@@ -0,0 +1,254 @@
+
+  proc unsignedVarintAppend(val:uint, ch: channel(true,iokind.little,false)) throws {
+    if val == 0 {
+      ch.write(b"\x00");
```
It is not a big deal but I think this is better written
```
var zero: uint(8);
ch.write(zero);
```

I understand that is longer (it's just that it keeps the code using simpler I/O).

Aniket21mathur replied:
Yeah, I will make the change :)

mppf commented on src/Encoding.chpl:
```diff
@@ -0,0 +1,254 @@
+  const fixed64Type = 1;
+  const fixed32Type = 5;
+
+  proc unsignedVarintAppend(val:uint, ch: channel(true,iokind.little,false)) throws {
```
It might be nice to use type aliases for this channel type, at the top of the file, like
```
type writingChannel = channel(true,iokind.little,false);
type readingChannel = channel(false,iokind.little,false);
```
this will reduce future maintenance work if these types change for some reason.

Aniket21mathur replied:
Yes! Agreed, will do that.

mppf commented on src/Encoding.chpl:
```diff
@@ -0,0 +1,254 @@
+  }
+
+  proc stringConsume(ch: channel(false,iokind.little,false)): string throws {
+    return bytesConsume(ch).decode();
```
This will throw if the string is not UTF-8. Is that expected behavior for protocol buffer messages? (If this is a complex question to answer, feel free to add a TODO to check in your task list).

Aniket21mathur replied:
> This will throw if the string is not UTF-8

Can you please give an example for which it throws, I tried with something like "~𝘈Ḇ𝖢𝕯٤ḞԍНǏ𝙅ƘԸⲘ𝙉০Ρ𝗤Ɍ𝓢ȚЦ", and it passes both general tests and end to end tests.

Aniket21mathur replied:
From the protobuf encoding guide for [Strings](https://developers.google.com/protocol-buffers/docs/encoding#strings)

> The red bytes are the UTF8 of "testing". The key here is 0x12 →

I understand that string type expects users to pass a UTF-8 value, other values can be handles through bytes. Though I am not 100% sure on this.

mppf replied:
I think protocol buffer strings are always UTF-8 - see https://developers.google.com/protocol-buffers/docs/proto3#scalar 

>   A string must always contain UTF-8 encoded or 7-bit ASCII text, and cannot be longer than 2^32.

Aniket21mathur replied:
Yes! Agreed.

mppf commented on src/Encoding.chpl:
```diff
@@ -0,0 +1,254 @@
+    const wireType = fixed32Type;
+    for i in 0..24 by 8 {
+      var newByte = (val >> i):uint(8);
+      ch.writeBytes(newByte, 1);
```
This can just be `ch.write(newByte)` but more generally this is just outputting the 32-bit value in big-endian format. Which makes me wonder if iokind.big would be a better setting for the channels? If iokind.big is the setting, then this would just be `ch.write(val)`.

This comment applies similarly to all of the fixed append and consume functions.

Aniket21mathur replied:
> This comment applies similarly to all of the fixed append and consume functions.

I am also a bit confused regarding the encoding/decoding of fixed types. The protobuf-encoding guide states them to be encoded as little-endian, but I saw a similar kind of implementation in Go protbuf library by Google(https://github.com/protocolbuffers/protobuf-go/blob/master/encoding/protowire/wire.go#L373).

Aniket21mathur replied:
> This can just be `ch.write(newByte)`

Ohh! I missed that, will change it :+1:

mppf replied:
> I am also a bit confused regarding the encoding/decoding of fixed types. The protobuf-encoding guide states them to be encoded as little-endian, but I saw a similar kind of implementation in Go protbuf library by Google(https://github.com/protocolbuffers/protobuf-go/blob/master/encoding/protowire/wire.go#L373).

That go implementation is writing it little endian. So is your loop here, I think I read it wrong before. Anyway you should be able to just write `ch.write(val)`. Give it a try?

Aniket21mathur replied:
Yes, it works!

mppf commented on src/plugin/message.cc:
```diff
@@ -62,24 +63,37 @@ namespace chapel {
+    printer->Indent(); 
+    
+    printer->Print(
+      "var binCh = new channel(writing=true, iokind.little, locking=false,\n"
```
I was expecting that this code would be in the user-facing Chapel module supporting protobuf. Anyway it would be better if we can find a way to avoid putting it in the protobuf compiler plugin because that makes it harder to update and it's relatively strange code.

Put another way, in an expected use case, where is writeToOutputFile called? Is it possible to handle the locking and making the binary channel at that point?

If not, is it possible to put these details in a helper function?

Aniket21mathur replied:
> Put another way, in an expected use case, where is writeToOutputFile called

`writeToOutputFile` is present in the chapel code generated through the plugin. It is called by the user in a `write.chpl` file (for e.g) to get the serialized values written to an `out` file. Example ->

```
use repeatedfield;
use IO;

var file = open("out", iomode.cw);
var writingChannel = file.writer();

var messageObj = new repeatedField();

messageObj.ui64.append(1);
messageObj.ui64.append(2445464);

messageObj.writeToOutputFile(writingChannel);
```

> I was expecting that this code would be in the user-facing Chapel module supporting protobuf.

I don't see a way of doing this with concrete functions, the `ch` channel from  `writeToOutputFile ` is passed directly to Append functions present in the user-facing Chapel module.

I also see that https://github.com/chapel-lang/chapel/pull/16032 is merged, can't we directly use the user exposed IO function for creating the new channel?

mppf replied:
> I also see that chapel-lang/chapel#16032 is merged, can't we directly use the user exposed IO function for creating the new channel?

Yes

> I don't see a way of doing this with concrete functions, the ch channel from writeToOutputFile is passed directly to Append functions present in the user-facing Chapel module.

Does that mean you do see a way of doing it with generic functions?

(e.g. message.writeToOutputFile calls Encoding.helpWriteToOutputFile(message, channel) which calls message.doWriteToOutputFile(channel) ?)

Aniket21mathur replied:
@mppf I made the changes in 382a749 and updated the output file in the pr message. Please review.

mppf replied:
The example output looks good. It will be relatively easy to add similar methods for reading from / writing to bytes using this mechanism!

Aniket21mathur replied:
Yes! The output is generated from the plugin itself. I think now we are good to go with this pr?

Aniket21mathur commented:
> I hope you address but you are welcome to list any of them as future work in your task list

Let's get as close as we can before merging the pr :)

mppf commented on src/Encoding.chpl:
```diff
@@ -0,0 +1,248 @@
+  proc tagConsume(ch: readingChannel) throws {
+    var (tag, tlen) = unsignedVarintConsume(ch);
+    if tlen == -1 then return -1;
+    var wireType = (tag & 0x7): int;
```
this and many other `var wireType` or `const wireTiype` are unused variables.

That might mean you can remove the `wireTypes` consts at the top, too.

Aniket21mathur replied:
Hmm..I would rather prefer keeping these for now, there are used when we ensure backward compatibility, though we can remove them anytime if they turn out to be useless.

mppf replied:
Sure, it makes sense to keep them in this PR if you plan to use them in a future PR.

mppf commented on src/Encoding.chpl:
```diff
@@ -0,0 +1,248 @@
+  const fixed64Type = 1;
+  const fixed32Type = 5;
+
+  proc unsignedVarintAppend(val:uint, ch: writingChannel) throws {
```
It's not something you have to change for this PR, but I wonder if `appendUnsignedVarint` would be a nicer method name. Not a big deal either way right now. (But, while these functions are not really intended to be user-facing, they do tie the protocol buffer compiler to particular versions of this module, so they should be stable). I recommend we revisit these names when we do the API review.

Aniket21mathur replied:
Yes, agreed. I have also opened an issue for the same(https://github.com/Aniket21mathur/Chapel-protobuf/issues/12).

mppf commented on src/Encoding.chpl:
```diff
@@ -0,0 +1,248 @@
+  }
+
+  proc writeToOutputFileHelper(ref message, ch) {
+    var binCh: channel(writing=true, kind=iokind.little, locking=false) = ch;
```
This should still lock and unlock the channel, right?

```
ch.lock();
defer { ch.unlock(); }
var binCh: channel(writing=true, kind=iokind.little, locking=false) = ch;
```

mppf commented on src/Encoding.chpl:
```diff
@@ -0,0 +1,248 @@
+  }
+
+  proc parseFromInputFileHelper(ref message, ch) {
+    var binCh: channel(writing=false, kind=iokind.little, locking=false) = ch;
```
needs to lock/unlock as well

mppf commented on src/Encoding.chpl:
```diff
@@ -0,0 +1,248 @@
+    return b;
+  }
+
+  proc writeToOutputFileHelper(ref message, ch) {
```
Shouldn't this function be marked `throws` ? As well as the read version and the code calling it generated by the protobuf compiler?

mppf commented on src/Encoding.chpl:
```diff
@@ -0,0 +1,248 @@
+/* Documentation for Encoding */
+module Encoding {
```
I'm not confident than Encoding will remain the name of this module but I recommend that we leave that choice for the API review. (In particular I think it probably should mention protobuf somewhere in the name).

Aniket21mathur commented:
@mppf Thanks for reviewing, I made the changes you suggested just leaving the function and module renaming for future API reviews.
